okhttp/src/main/java/okhttp3/Challenge.java

```diff
@@ -16,67 +16,92 @@
 import static okhttp3.internal.Util.ISO_8859_1;
 
 /** An RFC 7617 challenge. */
 public final class Challenge {
   private final String scheme;
-  private final String realm;
-  private final Charset charset;
+  private final Map<String, String> authParams;
 
-  public Challenge(String scheme, String realm) {
-    this(scheme, realm, ISO_8859_1);
+  public Challenge(String scheme, Map<String, String> authParams) {
+    if (scheme == null) throw new NullPointerException("scheme == null");
+    if (authParams == null) throw new NullPointerException("authParams == null");
+    this.scheme = scheme.toLowerCase();
+    Map<String, String> newAuthParams = new LinkedHashMap<>();
+    for (Entry<String, String> authParam : authParams.entrySet()) {
+      String key = (authParam.getKey() == null) ? null : authParam.getKey().toLowerCase();
+      newAuthParams.put(key, authParam.getValue());
+    }
+    this.authParams = unmodifiableMap(newAuthParams);
   }
 
-  private Challenge(String scheme, String realm, Charset charset) {
+  @Deprecated public Challenge(String scheme, String realm) {
     if (scheme == null) throw new NullPointerException("scheme == null");
     if (realm == null) throw new NullPointerException("realm == null");
-    if (charset == null) throw new NullPointerException("charset == null");
-    this.scheme = scheme;
-    this.realm = realm;
-    this.charset = charset;
+    this.scheme = scheme.toLowerCase();
+    this.authParams = singletonMap("realm", realm);
   }
 
   /** Returns a copy of this charset that expects a credential encoded with {@code charset}. */
-  public Challenge withCharset(Charset charset) {
-    return new Challenge(scheme, realm, charset);
+  @Deprecated public Challenge withCharset(Charset charset) {
+    if (charset == null) throw new NullPointerException("charset == null");
+    Map<String, String> authParams = new LinkedHashMap<>(this.authParams);
+    authParams.put("charset", charset.name());
+    return new Challenge(scheme, authParams);
   }
 
-  /** Returns the authentication scheme, like {@code Basic}. */
+  /** Returns the authentication scheme in lowercase, like {@code basic}. */
   public String scheme() {
     return scheme;
   }
 
+  /**
+   * Returns the auth params, including {@code realm} and {@code charset} if present,
+   * but as {@code String}s. The keys are all lowercase, as the auth param names
+   * are to be treated case insensitively anyway.
+   */
+  public Map<String, String> authParams() {
+    return authParams;
+  }
+
   /** Returns the protection space. */
   public String realm() {
-    return realm;
+    return authParams.get("realm");
   }
 
   /** Returns the charset that should be used to encode the credential. */
   public Charset charset() {
-    return charset;
+    String charset = authParams.get("charset");
+    if (charset != null) {
+      try {
+        return Charset.forName(charset);
+      } catch (Exception ignore) {
+      }
+    }
+    return ISO_8859_1;
```

Comment:
is this the official default? I’d prefer ASCII if possible

Reply:
Well, it is what OkHttp already had as default.
Actually it depends on the concrete auth scheme what is the default.
For `Basic` it is the default afair.
Actually I'd say the default should be `null` as it is auth scheme specific, but that will break implementations of `Basic` auth that depend it to be set.

My first thought was to have `Challenge` without any `getRealm` and `getCharset` methods as both are genrally optional and only mandatory by auth scheme and then have a `BasicChallenge` subclass that adds those methods for basic auth. But I tried to do my changes backwards compatibly.

What do you think?